rust/main/hyperlane-base/src/cache/moka/dynamic_expiry.rs

```diff
@@ -0,0 +1,83 @@
+use chrono::{offset::LocalResult, TimeZone, Utc};
+use moka::Expiry;
+use serde::{Deserialize, Serialize};
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+/// The type of expiration for a cache entry.
+///
+/// ## Variants
+///
+/// - `Never`: Never expire.
+/// - `AfterDuration`: Expire after a specified duration.
+/// - `AfterTimestamp`: Expire after a specified timestamp.
+/// - `Default`: Use the default expiration. (5 minutes)
+#[allow(missing_docs)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub enum ExpirationType {
+    Never,
+    AfterDuration(Duration),
+    AfterTimestamp(u64),
+    Default,
+}
+
+impl From<ExpirationType> for Expiration {
+    fn from(expiration: ExpirationType) -> Self {
+        Expiration {
+            variant: expiration,
+            created_at: Utc::now().timestamp(),
+        }
+    }
+}
+
+/// Expiration to store alongside a cache entry.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Expiration {
+    /// The type of expiration used when the entry was created.
+    pub variant: ExpirationType,
+    /// Unix timestamp when the entry was created.
+    pub created_at: i64,
+}
+
+impl Expiration {
+    /// Get the duration until the entry expires
+    pub fn as_duration(&self) -> Option<Duration> {
+        match self.variant {
+            ExpirationType::AfterDuration(duration) => Some(duration),
+            ExpirationType::AfterTimestamp(timestamp) => {
+                let target_time = UNIX_EPOCH + Duration::from_secs(timestamp);
+                target_time.duration_since(SystemTime::now()).ok()
+            }
```

Comment:
I wonder whether an entry will be expired if this method is called after the moment in time specified by `AfterTimestamp(timestamp)`?

Reply:
If the timestamp which is called with is in the past, the entity will auto expire, we can add a protection to allow only timestamps in the future

Reply:
I investigated further into this and you were correct, returning `None` if the target time is earlier than `now` would [cause the entry to never expire](https://docs.rs/moka/latest/moka/policy/trait.Expiry.html#method.expire_after_create).

A fix was implemented that if this were to happen, we would return a zero Duration, leading to instant expiry.
Also a test case was added for this case, as suggested.

[Fix](https://github.com/Mantas-M/hyperlane-monorepo/commit/c87b460e78a8f7f056348664a0578bf14ce6309f)

Good catch

Reply:
Nice.